src/lib.rs

```diff
@@ -41,6 +41,48 @@ enum OpenDALConfigurator {
     S3(opendal::services::S3Config),
 }
 
+#[derive(Debug)]
+struct SpeedTracker {
+    history: VecDeque<(Instant, u64)>,
+    window: Duration,
+    window_bytes: u64,
+}
+
+impl SpeedTracker {
+    fn new(window_secs: u64) -> Self {
+        Self {
+            history: VecDeque::new(),
+            window: Duration::from_secs(window_secs),
+            window_bytes: 0,
+        }
+    }
+
+    fn record(&mut self, bytes: u64, start_time: Instant) {
+        self.history.push_back((start_time, bytes));
+        self.window_bytes += bytes;
+
+        // remove entries older than the window
+        while let Some((time, bytes)) = self.history.front() {
+            if start_time.duration_since(*time) > self.window {
+                self.window_bytes -= bytes;
+                self.history.pop_front();
+            } else {
+                break;
+            }
+        }
+    }
+
+    fn avg_speed(&self) -> f64 {
+        if self.history.is_empty() {
+            return 0.0;
+        }
+
+        let duration = Instant::now().duration_since(self.history.front().unwrap().0);
+
+        (self.window_bytes as f64 / (1 << 20) as f64) / duration.as_secs_f64() //speed in MB/s
```

Comment:
we should also make it dynamic like we do with `HumanBytes`. Unfortunately indicatif doesn't provide a `HumanBits` implementation (speed is always provided as `Mb/s` or `Kb/s`) :( so I think the best way forward is copy over the `HumanBytes` implementation from indicatif into this project as `HumanBits` and modify it appropriately. I think we could even upstream this if they are interested in it.

Reply:
Fair, let's call it `HumanBitsPerSecond` though then